### PR TITLE
Add Python API to execute StableHLO bytecode

### PR DESCRIPTION
**Pull request by lsy323**
- Add a new Python API `_run_stablehlo` which takes StableHLO bytecode as well as input tensors. The StableHLO will be compiled by XLA and executed.

Example Usage
```
model = model.to(device)
inputs = tuple(i.to(device) for i in inputs if hasattr(i, 'to'))
output = model(*inputs)

bytecode = xm.get_stablehlo_bytecode([output])
res = torch_xla._XLAC._run_stablehlo(bytecode, list(inputs))
```

- With the new StableHLO execution API, the exported StableHLO program doesn't depend on TF for execution. The execution API is wrapped by the callable returned by the export function.

### CODE REVIEW

File: torch_xla/experimental/stablehlo_saved_model.py [hunk 1]
```diff
@@ -104,10 +108,12 @@ def export_torch_model(model: torch.nn.Module,
   args = tuple(
       tree_map_only(torch.Tensor, lambda x: x.to(device=device), sample_inputs))
   orig_state_dict = copy.copy(model.state_dict())
+  orig_state_dict = tree_map_only(torch.Tensor, lambda x: x.numpy(),
```
Comment (lsy323) [inline torch_xla/experimental/stablehlo_saved_model.py:111]: To align with the contract at https://github.com/pytorch/xla/blob/master/torch_xla/experimental/stablehlo_saved_model.py#L196, convert model weights to np.ndarray.

**Review by qihqi — APPROVED** [review]
few nits, overall good to go, thanks!

File: torch_xla/csrc/runtime/stablehlo_helper.cc [hunk 2]
```diff
@@ -105,5 +106,50 @@ std::string hloToStablehlo(const xla::HloModuleProto* proto,
...
+std::string GetHloModuleStr(const xla::HloModuleProto* proto) {
+  auto hlo_module = runtime::util::CreateModuleFromProto(*proto);
+  return hlo_module.value()->ToString();
+}
+
+static absl::Status stablehloToMhloHelper(mlir::ModuleOp* mlir_module,
+                                          mlir::MLIRContext* context) {
+  mlir::PassManager pm(context);
+  pm.addPass(mlir::mhlo::createStablehloLegalizeToHloPass());
+  if (!mlir::succeeded(pm.run(*mlir_module))) {
```
[the remark this answers is not in the document]
Comment (lsy323) [reply torch_xla/csrc/runtime/stablehlo_helper.cc:118]: Yes, it's a MLIR pass that does stablehlo -> mhlo

File: torch_xla/csrc/runtime/stablehlo_helper.cc [hunk 3]
```diff
@@ -105,5 +106,50 @@ std::string hloToStablehlo(const xla::HloModuleProto* proto,
   }
 }
 
+std::string GetHloModuleStr(const xla::HloModuleProto* proto) {
+  auto hlo_module = runtime::util::CreateModuleFromProto(*proto);
+  return hlo_module.value()->ToString();
+}
+
+static absl::Status stablehloToMhloHelper(mlir::ModuleOp* mlir_module,
```
[the remark this answers is not in the document]
Comment (lsy323) [reply torch_xla/csrc/runtime/stablehlo_helper.cc:114]: sg, will update

File: torch_xla/csrc/runtime/stablehlo_helper.cc [hunk 4]
```diff
@@ -105,5 +106,50 @@ std::string hloToStablehlo(const xla::HloModuleProto* proto,
...
+  pm.addPass(mlir::mhlo::createStablehloLegalizeToHloPass());
+  if (!mlir::succeeded(pm.run(*mlir_module))) {
+    return absl::Status(
+        absl::StatusCode::kInternal,
+        "StableHLO Module from StableHLO -> MHLO conversion is not leagal.");
+  }
+  return absl::OkStatus();
+}
+
+static absl::Status MhloToHloHelper(const mlir::ModuleOp* mlir_module,
```
[the remark this answers is not in the document]
Comment (lsy323) [reply torch_xla/csrc/runtime/stablehlo_helper.cc:126]: sg, will update

File: torch_xla/csrc/init_python_bindings.cpp [hunk 5]
```diff
@@ -1187,6 +1187,23 @@ void InitXlaModuleBindings(py::module m) {
...
+        [](const std::string& bytecode,
+           const std::vector<at::IValue>& graph_inputs)
+            -> std::vector<at::Tensor> {
+          torch::lazy::BackendDevice device = torch_xla::GetCurrentDevice();
+          auto results = XLAGraphExecutor::Get()->ExecuteStablehlo(
+              bytecode, graph_inputs, device);
+          std::vector<at::Tensor> retlist;
+          {
+            // Convert result back to at::tensor
+            for (auto& data : results) {
```
[the remark this answers is not in the document]
Comment (lsy323) [reply torch_xla/csrc/init_python_bindings.cpp:1200]: Thanks, will update

File: torch_xla/csrc/runtime/stablehlo_helper.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (JackCaoG) [reply torch_xla/csrc/runtime/stablehlo_helper.cc]: I am fine with the helper but I think you should capitalize the stable

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 6]
```diff
@@ -755,6 +757,80 @@ XLAGraphExecutor::ExecuteComputationWithBarrier(
...
+          runtime::GetComputationClient()->Compile(std::move(instances));
+
+  // Placeholder for computation outputs.
+  std::vector<torch::lazy::BackendDataPtr> placeholders;
+  for (const auto& s : flatten_shapes) {
+    torch::lazy::BackendDataPtr handle =
+        WrapXlaData(runtime::GetComputationClient()->CreateDataPlaceholder(
+            device.toString(), std::move(s)));
+    placeholders.push_back(handle);
+  }
```
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:805]: you don't really need placeholders if the computation happens on the same thread. The purpose of placeholder is this function can return and the actual compuations happens in a background thread.

Comment (lsy323) [reply torch_xla/csrc/xla_graph_executor.cpp]: Thanks Jack, updated

File: torch_xla/csrc/xla_graph_executor.cpp [hunk 7]
```diff
@@ -824,9 +809,9 @@ std::vector<torch::lazy::BackendDataPtr> XLAGraphExecutor::ExecuteStablehlo(
       runtime::GetComputationClient()->ExecuteComputation(
           *computations[0], UnwrapXlaData(arguments), device.toString());
 
+  std::vector<torch::lazy::BackendDataPtr> placeholders;
   for (size_t i = 0; i < results.size(); ++i) {
-    XLA_CHECK(placeholders[i] != nullptr);
-    placeholders[i]->Assign(*WrapXlaData(results[i]));
+    placeholders.push_back(WrapXlaData(results[i]));
```
Comment (JackCaoG) [inline torch_xla/csrc/xla_graph_executor.cpp:814]: nit you should not call them placeholders, you can just call it `results` or `datas`

**Review by JackCaoG — APPROVED** [review]

**Pull request merged**